tools/myst/enc/syscall.c

```diff
@@ -1637,6 +1603,132 @@ static long _fsync(int fd)
 }
 #endif
 
+static int _pipe2(int pipefd[2], int flags)
+{
+    int ret = 0;
+    long retval;
+
+    if (!pipefd)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    if (myst_pipe2_ocall(&retval, pipefd, flags) != OE_OK)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    ret = retval;
+
+done:
+    return ret;
+}
+
+static long _epoll_create1(int flags)
+{
+    long ret = 0;
+    long retval;
+
+    if (myst_epoll_create1_ocall(&retval, flags) != OE_OK)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    ret = retval;
+
+done:
+    return ret;
+}
+
+static long _epoll_wait(
+    int epfd,
+    struct epoll_event* events,
+    size_t maxevents,
+    int timeout)
+{
+    long ret = 0;
+    long retval;
+
+    if (epfd < 0 || !events || maxevents == 0)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    if (myst_epoll_wait_ocall(&retval, epfd, events, maxevents, timeout) !=
+        OE_OK)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    /* guard against possible read past end of buffer */
+    if (retval > maxevents)
```

Comment:
Encountered a failure here with a .NET coreclr test case. Looks like the compiler converted int retval to size_t before the comparison and caused negative retval, such a -EINTR, to trigger the error condition.

```suggestion
    if ((retval > 0) && (retval > maxevents))
```

Reply:
Thank you Bo for finding this. The comparison between `retval` and `maxevents` should have reported a conversion error but unfortunately that compiler option is not enabled. We need to enable checks for conversions across Mystikos but for now, I have applied the following fix.

```C++
    /* check for error */
    if (retval < 0)
    {
        ret = retval;
        goto done;
    }

    /* guard against possible read past end of buffer */
    if ((size_t)retval > maxevents)
    {
        ret = -EINVAL;
        goto done;
    }
```